File: src/libs/jderobotcomm_cpp/src/ros/publisherMotors.cpp
#include <jderobot/comm/ros/publisherMotors.hpp>

namespace JdeRobotComm {

	PublisherMotors::PublisherMotors(int argc, char** argv, std::string nodeName, std::string topic){
		pthread_mutex_init(&mutex, NULL);
		if ("" == topic){
			this->on = false;
			std::cerr <<"Invalid Motors topic" <<std::endl;
		}else{
			this->on = true;
			this->topic = topic;
			this->nodeName = nodeName;
			//boost::thread t3(&PublisherMotors::listen, this);

			const std::string name = std::string(this->nodeName);
			int a = 0;
			ros::init(a, nullptr, name);
			ros::NodeHandle nh;
			this->pub = nh.advertise<geometry_msgs::Twist>(this->topic, 1, true);
			std::cout << "publishing in "+ this->topic << std::endl;

			timer_ = nh.createTimer(ros::Duration(0.1), boost::bind(&PublisherMotors::publish, this));

			this->spinner = new ros::AsyncSpinner(1);
			
		}
	}


	void PublisherMotors::publish()
	{
		geometry_msgs::Twist vel;
		JdeRobotTypes::CMDVel cmdvel;
		pthread_mutex_lock(&mutex);
			cmdvel = this->cmdvel;
		pthread_mutex_unlock(&mutex);
		vel = translate_twist_messages (cmdvel);
		this->pub.publish(vel);


	}



	PublisherMotors::~PublisherMotors(){
		this->stop();
	}

	void 
	PublisherMotors::start(){
		this->spinner->start();
	}

	void 
	PublisherMotors::stop(){
		this->spinner->stop();
		ros::shutdown();
	}

	void
	PublisherMotors::sendVelocities(JdeRobotTypes::CMDVel vel){
		pthread_mutex_lock(&mutex);
			this->cmdvel = vel;
		pthread_mutex_unlock(&mutex);
	}

	void
	PublisherMotors::sendVX (float vx){
		pthread_mutex_lock(&mutex);
			this->cmdvel.vx = vx;
		pthread_mutex_unlock(&mutex);
	}

	void
	PublisherMotors::sendVY (float vy){
		pthread_mutex_lock(&mutex);
			this->cmdvel.vy = vy;
		pthread_mutex_unlock(&mutex);

	}

	void
	PublisherMotors::sendAZ (float az){
		pthread_mutex_lock(&mutex);
			this->cmdvel.az = az;
		pthread_mutex_unlock(&mutex);
	}

	void
	PublisherMotors::sendV (float v){
		this->sendVX(v);
	}

	void
	PublisherMotors::sendW (float w){
		this->sendAZ(w);
	}

	void
	PublisherMotors::sendL (float l ){
		this->sendVY(l);
	}



}//NS

File: src/libs/comm_py/comm/ros/publisherMotors.py
import rospy
from geometry_msgs.msg import Twist
import threading
from math import pi as PI
from jderobotTypes import CMDVel
from .threadPublisher import ThreadPublisher



def cmdvel2Twist(vel):
    '''
    Translates from JderobotTypes CMDVel to ROS Twist. 

    @param vel: JderobotTypes CMDVel to translate

    @type img: JdeRobotTypes.CMDVel

    @return a Twist translated from vel

    '''
    tw = Twist()
    tw.linear.x = vel.vx
    tw.linear.y = vel.vy
    tw.linear.z = vel.vz
    tw.angular.x = vel.ax
    tw.angular.y = vel.ay
    tw.angular.z = vel.az

    return tw

class PublisherMotors:
    '''
        ROS Motors Publisher. Motors Client to Send CMDVel to ROS nodes.
    '''
    def __init__(self, topic, maxV, maxW):
        '''
        ListenerMotors Constructor.

        @param topic: ROS topic to publish
        
        @type topic: String

        '''
        self.maxW = maxW
        self.maxV = maxV

        self.topic = topic
        self.data = CMDVel()
        self.pub = self.pub = rospy.Publisher(self.topic, Twist, queue_size=1)
        self.lock = threading.Lock()

        self.kill_event = threading.Event()
        self.thread = ThreadPublisher(self, self.kill_event)

        self.thread.daemon = True
        self.start()
 
    def publish (self):
        '''
        Function to publish cmdvel. 
        '''
        self.lock.acquire()
        tw = cmdvel2Twist(self.data)
        self.lock.release()
        self.pub.publish(tw)
        
    def stop(self):
        '''
        Stops (Unregisters) the client. If client is stopped you can not start again, Threading.Thread raised error

        '''
        self.kill_event.set()
        self.pub.unregister()

    def start (self):
        '''
        Starts (Subscribes) the client. If client is stopped you can not start again, Threading.Thread raised error

        '''
        self.kill_event.clear()
        self.thread.start()
        


    def getMaxW(self):
        return self.maxW

    def getMaxV(self):
        return self.maxV
        

    def sendVelocities(self, vel):
        '''
        Sends CMDVel.

        @param vel: CMDVel to publish
        
        @type vel: CMDVel

        '''
        self.lock.acquire()
        self.data = vel
        self.lock.release()

    def sendV(self, v):
        '''
        Sends V velocity. uses self.sendVX

        @param v: V velocity
        
        @type v: float

        '''
        self.sendVX(v)

    def sendL(self, l):
        '''
        Sends L velocity. uses self.sendVY

        @param l: L velocity
        
        @type l: float

        '''
        self.sendVY(l)

    def sendW(self, w):
        '''
        Sends W velocity. uses self.sendAZ

        @param w: W velocity
        
        @type w: float

        '''
        self.sendAZ(w)

    def sendVX(self, vx):
        '''
        Sends VX velocity.

        @param vx: VX velocity
        
        @type vx: float

        '''
        self.lock.acquire()
        self.data.vx = vx
        self.lock.release()

    def sendVY(self, vy):
        '''
        Sends VY velocity.

        @param vy: VY velocity
        
        @type vy: float

        '''
        self.lock.acquire()
        self.data.vy = vy
        self.lock.release()

    def sendAZ(self, az):
        '''
        Sends AZ velocity.

        @param az: AZ velocity
        
        @type az: float

        '''
        self.lock.acquire()
        self.data.az = az
        self.lock.release()


